rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+- Start Date: 2020-04-19
+- RFC PR: [#64284](https://github.com/elastic/kibana/pull/64284)
+- Kibana Issue: [#61657](https://github.com/elastic/kibana/issues/61657)
+
+# Summary
+
+A core API exposed on both public and server sides, allowing to search for various objects and
+register result providers.
+
+# Basic example
+
+- registering a result provider:
+
+```ts
+coreSetup.globalSearch.registerResultProvider({
+  id: 'my_provider',
+  find: (term, options, context) => {
+    const resultPromise = myService.search(term, context.core.savedObjects.client);
+    return from(resultPromise);
+  },
+});
+```
+
+- using the `find` API from the client-side:
+
+```ts
+coreStart.globalSearch.find('some term').subscribe(({ results, complete }) => {
+    showAsyncSearchIndicator(!complete);
+    updateResults(results);
+});
+```
+
+# Motivation
+
+Kibana should do it's best to assist users searching for and navigating to the various objects present on the Kibana platform.
+
+We should expose an API to make it possible for plugins to search for the various objects present on a Kibana instance.
+
+The first consumer of this API will be the global search bar [#57576](https://github.com/elastic/kibana/issues/57576). This API
+should still be generic to answer similar needs from any other consumer, either client or server side.
+
+# Detailed design
+
+## API Design
+
+### Result provider API
+
+#### common types
+
+```ts
+/**
+ * Static list of all the possible result types.
+ * Ordinal value matters here, as it will be used to sort results of different types.
+ */
+enum SEARCH_TYPE {
+  // non-exhaustive
+  applications = 10,
+  dashboard = 20,
+  visualization = 30,
+  search = 40,
+}
+
+/** @public */
+type GlobalSearchResultType = keyof typeof SEARCH_TYPE;
+
+/**
+ * Options provided to {@link GlobalSearchResultProvider | result providers} `find` method.
+ * Currently empty and only present to keep the API future-proof.
+ */
+interface GlobalSearchOptions {}
+
+/**
+ * Representation of a result returned by a {@linlkGlobalSearchResultProvider | result provider}
+ */
+interface GlobalSearchResult {
+  /** an id that should be unique for an individual provider's results */
+  id: string;
+  /** the title/label of the result */
+  title: string;
+  /** the type of result */
+  type: GlobalSearchResultType;
```

Comment:
@elastic/kibana-core-ui  Do you see any missing field here?

Reply:
This list looks correct. Two possibly related notes:
- The icon will only display for applications (e.g. Dashboard, Discover, Logs, etc.) not the objects. The comments notes a default icon would be used, but it would actually just not display any icon at all.
![image](https://user-images.githubusercontent.com/446285/80222561-aa1d5500-860c-11ea-9911-26dd43f819e2.png)


- In the default state (upon initial open), we would like to display recent items in the results list until you enter a search string. Per the design, we would show an icon to denote this as a 'recent' item as opposed to a standard search result item. I don't think this affects the API, so I mention it here just to raise awareness.

![image](https://user-images.githubusercontent.com/446285/80222414-69bdd700-860c-11ea-8d02-5fdd4ec4e852.png)

Reply:
> The icon will only display for applications (e.g. Dashboard, Discover, Logs, etc.) not the objects

What the search bar decides to do based on the type of result is outside of the scope of the GS API. However this is not an issue per say, as choosing to display the icon only for a given type should be trivial for the search bar implementation.

> The comments notes a default icon would be used, but it would actually just not display any icon at all.

This is a good remark, This behavior is outside of the scope of the GS API. What a consumer choose to do when an icon is not present is an implementation detail.
Will remove the ` If not specified, will use a default icon` part.

> In the default state (upon initial open), we would like to display recent items in the results list until you enter a search string

This is out of the scope of the globalSearch API in it's current specifications. 'recent items' feature should be managed by the search bar component itself.